### PR TITLE
Add C# namespace to .proto files (e.g. descriptors, unit tests)

### PR DESCRIPTION
**Pull request by jskeet**

### CODE REVIEW

**Comment by jskeet** [comment]
@jtattermusch and @anandolee to review

File: benchmarks/google_size.proto [hunk 1]
```diff
@@ -1,7 +1,10 @@
+syntax = "proto2";
```
Comment (jtattermusch) [inline benchmarks/google_size.proto]: why do you need this?

File: src/google/protobuf/unittest_import_public.proto [hunk 2]
```diff
@@ -35,6 +35,7 @@ syntax = "proto2";
 package protobuf_unittest_import;
 
 option java_package = "com.google.protobuf.test";
+option csharp_namespace = "Google.ProtocolBuffers.TestProtos";
```
Comment (jtattermusch) [inline src/google/protobuf/unittest_import_public.proto]: It makes me wonder why for some test protos, java uses 
com.google.protobuf.test, for others it uses com.google.protobuf and for some it doesn't define the package at all. Do we really need csharp_namespace for all the test protos? It's for tests, rights, so arguably we don't care much what namespace we end up with with the generated protos.
Feel free to ignore this comment.

**Comment by jtattermusch** [comment]
Generally, it looks fine, but this is more for @anandolee to decide.

File: benchmarks/google_size.proto (hunk 1 above)
Comment (jskeet) [inline benchmarks/google_size.proto]: Without it, protoc issues a warning. Given that it *is* a proto2 file, it felt reasonable to be explicit. I was somewhat surprised that it was still unspecified, to be honest...

File: src/google/protobuf/unittest_import_public.proto (hunk 2 above)
Comment (jskeet) [inline src/google/protobuf/unittest_import_public.proto]: Yes, I wondered about that. It ends up looking "odd" to have unconventional namespaces in the test code - I don't have much opinion either way though. If we don't want this, I'll create another commit which changes the namespace in the generated (but present-in-git) code.

**Comment by jskeet** [comment]
Just checked the Travis CI failure, and it's because descriptor.proto has changed but I haven't updated the other files. I'll need to regenerate that on a Linux box... will make the changes in another commit and push that. (It may well be tomorrow...)

**Comment by xfxyjwf** [comment]
Do we have such a "csharp_namespace" option in our internal code base? If not I suggest we add one first. The .proto files modified in this PR is pulled from our internal code base. We will need to merge all changes here back to google3.

**Comment by jskeet** [comment]
I'm assuming that we'll merge the entire branch back at some point - I don't know whether *any* of the C# work has been merged back yet...

File: src/google/protobuf/descriptor.proto [hunk 3]
```diff
@@ -43,6 +43,8 @@ package google.protobuf;
 option java_package = "com.google.protobuf";
 option java_outer_classname = "DescriptorProtos";
```
Comment (anandolee) [inline src/google/protobuf/descriptor.proto]: remove empty line

Comment (jskeet) [inline src/google/protobuf/descriptor.proto]: Done. Wasn't sure whether we wanted to keep the C# options separate from the Java ones. Will check the other ones in the commit...

**Comment by anandolee** [comment]
I think our current plan is not support c# internally (unless we find some reason to support it), so we don't want to expose csharp options in internal code as well. Discussed with Feng that we should add the ability to remove these options for up integrade.  After that, there should be no problem for this Change. 
Just make sure to fix the Travis CI failure.

**Comment by jskeet** [comment]
@anandolee Just to check, are you happy for this change to be pulled into this branch (when Travis is happy, which it should be now... just waiting for confirmation) and we can fix up things internally before the csharp branch is merged into the master branch?

**Comment by anandolee** [comment]
Yes, go ahead to pull into this branch. I will take care about the internal change before merge into internal code.
LGTM

**Pull request merged**